### PR TITLE
Remove ScalarFunctionDefinition

### PR DESCRIPTION
**Pull request by lewiszlw**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion/issues/10175.

### CODE REVIEW

File: datafusion/expr/src/expr_schema.rs [hunk 1]
```diff
@@ -138,25 +138,21 @@ impl ExprSchemable for Expr {
                     .iter()
                     .map(|e| e.get_type(schema))
                     .collect::<Result<Vec<_>>>()?;
-                match func_def {
-                    ScalarFunctionDefinition::UDF(fun) => {
                         // verify that function is invoked with correct number and type of arguments as defined in `TypeSignature`
-                        data_types(&arg_data_types, fun.signature()).map_err(|_| {
+                        data_types(&arg_data_types, func_def.signature()).map_err(|_| {
```
Comment (jayzhan211) [inline datafusion/expr/src/expr_schema.rs:142]: I think it is fine to name it `func`

**Review by jayzhan211 — APPROVED** [review]
👍

**Review by Weijun-H — APPROVED** [review]
LGTM! Thanks @lewiszlw

**Comment by alamb** [comment]
🚀

File: datafusion/expr/src/expr.rs [hunk 2]
```diff
@@ -399,18 +399,11 @@ impl Between {
...
-    /// Resolved to a user defined function
-    UDF(Arc<crate::ScalarUDF>),
-}
-
 /// ScalarFunction expression invokes a built-in scalar function
 #[derive(Clone, PartialEq, Eq, Hash, Debug)]
 pub struct ScalarFunction {
     /// The function
-    pub func_def: ScalarFunctionDefinition,
+    pub func_def: Arc<crate::ScalarUDF>,
```
Comment (alamb) [inline datafusion/expr/src/expr.rs:406]: If we are changing the API, do we want to change this to `func` too while we are at it?

**Review by alamb — APPROVED** [review]
Thanks @lewiszlw

**Comment by alamb** [comment]
It might be nice to wait until we make the #10217  release, to minimize API churn as 38.0.0  already removes BuiltInScalarFunction

File: datafusion/expr/src/expr.rs (hunk 2 above)
Comment (lewiszlw) [reply datafusion/expr/src/expr.rs]: Agree. Changed.

**Comment by lewiszlw** [comment]
Clippy failed on other code. I run clippy locally and more than one warning needed to be addressed, might be done in another pr.

**Comment by alamb** [comment]
I merged up from main to pick up the fix for https://github.com/apache/datafusion/issues/10352

**Comment by alamb** [comment]
Now that https://github.com/apache/datafusion/pull/10407 is merged, I believe main is open for 38.0.0 business !

🚀  -- tanks again @lewiszlw

**Pull request merged**